barcode-manager/man/bmcli.1.in
.\" bmcli (1) manual page
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111,
.\" USA.
.\"
.\" Copyright (C) 2011 Jakob Flierl
.\"
.TH NMCLI "1" "26 November 2011"

.SH NAME
bmcli \- command-line tool for controlling BarcodeManager
.SH SYNOPSIS
.ad l
.B bmcli
.RI " [ " OPTIONS " ] " OBJECT " { " COMMAND " | "
.BR help " } "
.sp

.IR OBJECT " := { "
.BR bm " | " con " | " dev " } "
.sp

.IR OPTIONS " := { "
.br
\fB\-t\fR[\fIerse\fR]
.br
\fB\-p\fR[\fIretty\fR]
.br
\fB\-m\fR[\fImode\fR] tabular | multiline
.br
\fB\-f\fR[\fIields\fR] <field1,field2,...> | all | common
.br
\fB\-e\fR[\fIscape\fR] yes | no
.br
\fB\-v\fR[\fIersion\fR]
.br
\fB\-h\fR[\fIelp\fR]
.br
.RI "}"

.SH DESCRIPTION
.B bmcli
is a command-line tool for controlling BarcodeManager and getting its status.
It is not meant as a replacement of \fIbm-applet\fP or other similar clients.
Rather it's a complementary utility to these programs.
The main \fIbmcli\fP's usage is on servers, headless machines or just for
power users who prefer the command line.
.P
The use cases comprise:
.IP \(em 4
Initscripts: can utilize BarcodeManager via \fIbmcli\fP instead of
having to manage barcode scanners itself and possible interfere with BarcodeManager.
.IP \(em 4
Servers, headless machines: No GUI is available; then \fIbmcli\fP is used to
talk directly to BarcodeManager and control only system-wide barcode scanners.
.IP \(em 4
User sessions: For this case, \fIbmcli\fP can talk to \fIbm-applet\fP to find
user confogured barcode scanners. It can still talk directly to BarcodeManager for
manipulating these barcode scanners. As \fIbmcli\fP doesn't have direct access to
user configuration data in GConf, \fIbm-applet\fP handles that itself. That may,
for example, cause the applet to pop up dialogs when user input is needed.
.SS \fIOPTIONS\fP
.TP
.B \-t, \-\-terse
Output is terse.  This mode is designed and suitable for computer (script)
processing.
.TP
.B \-p, \-\-pretty
Output is pretty.  This causes \fIbmcli\fP to produce easy readable outputs
for humans, i.e. values are aligned, headers are printed, etc.
.TP
.B \-m, \-\-mode tabular | multiline
Switch between \fItabular\fP and \fImultiline\fP output.
If omitted, default is \fItabular\fP for most commands.  For the commands
producing more structured information, that cannot be displayed on a single
line, default is \fImultiline\fP. Currenly, they are:
.br
.nf
  'bmcli con list id|uuid <name>'
  'bmcli dev list'
.fi
\fItabular\fP   - Output is a table where each line describes a single entry.
Columns define particular properties of the entry.
.br
\fImultiline\fP - Each entry comprises more lines, each property on its own line.
The values are prefixed with the property name.
.TP
.B \-f, \-\-fields <field1,field2,...> | all | common
This option is used to specify what fields (column names) should be printed.
Valid field names differ for specific commands. List available fields by
providing an invalid value to the \fI--fields\fP option.
.br
\fIall\fP is used to print all valid field values of the command.
\fIcommon\fP is used to print common field values of the command.
If omitted, default is \fIcommon\fP.
The option is mandatory when \fI--terse\fP is used.  In this case, generic
values \fIall\fP and \fIcommon\fP cannot be used.  (This is to maintain
compatibility when new fields are added in the future).
.TP
.B \-e, \-\-escape yes | no
Whether to escape ':' and '\\' characters in terse tabular mode.  The escape
character is '\\'.
If omitted, default is \fIyes\fP.
.TP
.B \-v, \-\-version
Show \fIbmcli\fP version.
.TP
.B \-h, \-\-help
Print help information.
.SS \fIOBJECT\fP
.TP
.B bm
BarcodeManager
.br
Use this object to inquire and change state of BarcodeManager.
.TP
.SS \fICOMMAND\fP := { status | enable | sleep }
.sp
.RS
.TP
.B status
.br
Show overall status of BarcodeManager.  This is the default action, when no
command is provided to \fIbm\fP object.
.br
.nf
\fBReference to D-Bus:\fP
No simple reference.
.fi
.TP
.B enable [true|false]
.br
Get status or enable barcode scanners by BarcodeManager.
All barcode scanners managed by BarcodeManager are deactivated when disabled.
.br
.nf
\fBReference to D-Bus:\fP
interface: org.freedesktop.BarcodeManager
method:    Enable
arguments: TRUE or FALSE
.fi
.TP
.B sleep [true|false]
.br
Get sleep status or put to sleep/awake BarcodeManager. All barcode scanners
managed by BarcodeManager are deactivated when it falls asleep.  This command is
not meant for user to enable/disable barcode scanners, use \fIenable\fP for that.
D-Bus \fISleep\fP method is designed to put BarcodeManager to sleep or awake for
suspending/resuming computer.
.br
.nf
\fBReference to D-Bus:\fP
interface: org.freedesktop.BarcodeManager
method:    Sleep
arguments: TRUE or FALSE
.fi
.TP
.B wifi [on|off]
.br
Inquire or set status of WiFi connected scanners in BarcodeManager.Without any
further argument, WiFi status is printed; \fIon\fP enables WiFi based barcode
scanners; \fIoff\fP disables barcode scanners WiFi.
.br
.nf
\fBReference to D-Bus:\fP
No simple reference.
.fi
.TP
.B wwan [on|off]
.br
Inquire or set status of WWAN in BarcodeManager.  Without any further argument,
WWAN status is printed; \fIon\fP enables WWAN; \fIoff\fP disables WWAN.
.br
.nf
\fBReference to D-Bus:\fP
No simple reference.
.fi
.RE

.TP
.B con
Connections
.br
Get information about BarcodeManager's connections.
.TP
.SS \fICOMMAND\fP := { list | status | up | down }
.sp
.RS
.TP
.B list [id <id> | uuid <id> | system | user]
.br
List configured connections.  Without a parameter, configured connection from
both system and user settings services are listed.  \fIsystem\fP argument filters
only system-wide connections, \fIuser\fP prints user connections only.
In order to get connection details, \fIid\fP with connection's name or \fIuuid\fP
with connection's UUID shall be specified.
When no command is given to \fIcon\fP object, the default action is 'nmcli con list'.
.br
.nf
\fBReference to D-Bus:\fP
No simple reference.
.fi
.TP
.B status
.br
Print status of active connections.
.br
.nf
\fBReference to D-Bus:\fP
No simple reference.
.fi
.TP
.B up id <id> | uuid <id> [iface <iface>] [ap <hwaddr>] [\-\-nowait] [\-\-timeout <timeout>]
.br
Activate a connection.  The connection is identified by its name using \fIid\fP
or UUID using \fIuuid\fP.  For requiring particular device to activate
the connection on, \fIiface\fP option with interface name should be given.
\fIap\fP option can further concretize what AP should be used in case of WiFi
connection.  \fI\-\-nowait\fP option causes \fInmcli\fP to exit immediately and
not to wait for command completion.  \fI\-\-timeout\fP option provides a means
to specify how long to wait for operation completion.
.br
.nf
\fBReference to D-Bus:\fP
interface: org.freedesktop.BarcodeManager
method:    ActivateConnection
arguments: according to arguments
.fi
.TP
.B down id <id> | uuid <id>
.br
Deactivate a connection.
The connection is identified by its name using \fIid\fP
or UUID using \fIuuid\fP.
.br
.nf
\fBReference to D-Bus:\fP
interface: org.freedesktop.BarcodeManager
method:    DeactivateConnection
arguments: according to arguments
.fi
.RE

.TP
.B dev
Devices
.br
Get information about devices.
.TP
.SS \fICOMMAND\fP := { status | list | disconnect | wifi }
.sp
.RS
.TP
.B status
.br
Print status of devices.  This is the default action, when no command
is specified to \fIdev\fP object.
.br
.nf
\fBReference to D-Bus:\fP
No simple reference.
.fi
.TP
.B list [iface <iface>]
.br
Get detailed information about devices.  Without an argument, all devices are
examined.  To get information for a specific device, \fIiface\fP argument
with the interface name should be provided.
.br
.nf
\fBReference to D-Bus:\fP
No simple reference.
.fi
.TP
.B disconnect iface <iface> [\-\-nowait] [\-\-timeout <timeout>]
.br
Disconnect a device and prevent the device from automatically activating further
connections without user/manual intervention.  \fI\-\-nowait\fP option causes
\fInmcli\fP to exit immediately and not to wait for command completion.
\fI\-\-timeout\fP option provides a means to specify how long to wait for
operation completion.
.br
.nf
\fBReference to D-Bus:\fP
interface: org.freedesktop.BarcodeManager.Device
method:    Disconnect
arguments: none
.fi
.TP
.B wifi [list [iface <iface>] [hwaddr <hwaddr>]]
.br
List available WiFi access points.  \fIiface\fP and \fIhwaddr\fP options
can be used to get just APs for particular interface or specific AP,
respectively.
.br
.nf
\fBReference to D-Bus:\fP
No simple reference.
.fi
.RE

.SH ENVIRONMENT VARIABLES
\fInmcli\fP's behavior is affected by the following environment variables.
.IP "LC_ALL" 13
If set to a non-empty string value, override the values of all the other
internationalization variables.
.IP "LC_MESSAGES" 13
Determines the locale to be used for internationalised messages.
.IP "LANG" 13
Provides a default value for the internationalization variables that are unset
or null.

.RE
Notes about localization:
.br
Be aware that \fInmcli\fP is localized and that's why the output depends on 
your environment. It's important to realize that especially when you parse the
output.
.br
Call \fInmcli\fP as \fBLC_ALL=C nmcli\fP to be sure the locale is
set to "C" while executing in a script.

\fBLC_ALL\fP, \fBLC_MESSAGES\fP, \fBLANG\fP variables specify the LC_MESSAGES
locale category (in that order), which determines the language that \fInmcli\fP
uses for messages.  The "C" locale is used if none of these variables are set,
and this locale uses English messages.

.SH EXIT STATUS
\fInmcli\fP exits with status 0 if it succeeds, a value greater than 0 is returned if errors occur.
.IP "0" 4
Success - indicates the operation succeeded
.IP "1" 4
Unknown or unspecified error
.IP "2" 4
Invalid user input, wrong \fInmcli\fP invocation
.IP "3" 4
Timeout expired (see commands with \fI\-\-timeout\fP option)
.IP "4" 4
Connection activation failed
.IP "5" 4
Connection deactivation failed
.IP "6" 4
Disconnecting device failed

.SH BUGS
There are probably some.  If you find a bug, please report to
https://github.com/koppi/barcode-utils

.SH SEE ALSO
.BR bm\-tool (1),
.BR bm\-online (1),
.BR BarcodeManager(8).


barcode-manager/cli/src/bmcli.c
/* bmcli - command-line tool to control BarcodeManager
 *
 * Jakob Flierl <jakob flierl@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * (C) Copyright 2011 Jakob Flierl
 */

/* Generated configuration file */
#include "config.h"

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <signal.h>
#include <locale.h>

#include <glib.h>
#include <glib/gi18n.h>
#include <bm-client.h>
#include <bm-setting-connection.h>
#include <bm-remote-settings.h>
#include <bm-remote-settings-system.h>
#include <bm-settings-interface.h>
#include <bm-settings-connection-interface.h>

#include "bmcli.h"
#include "utils.h"
#include "connections.h"
#include "devices.h"
#include "barcode-manager.h"

#if defined(BM_DIST_VERSION)
# define BMCLI_VERSION BM_DIST_VERSION
#else
# define BMCLI_VERSION VERSION
#endif


typedef struct {
        BmCli *bmc;
	int argc;
	char **argv;
} ArgsInfo;

/* --- Global variables --- */
GMainLoop *loop = NULL;


static void
usage (const char *prog_name)
{
	fprintf (stderr,
	         _("Usage: %s [OPTIONS] OBJECT { COMMAND | help }\n\n"
	         "OPTIONS\n"
	         "  -t[erse]                                   terse output\n"
	         "  -p[retty]                                  pretty output\n"
	         "  -m[ode] tabular|multiline                  output mode\n"
	         "  -f[ields] <field1,field2,...>|all|common   specify fields to output\n"
	         "  -e[scape] yes|no                           escape columns separators in values\n"
	         "  -v[ersion]                                 show program version\n"
	         "  -h[elp]                                    print this help\n\n"
	         "OBJECT\n"
	         "  bm          BarcodeManager status\n"
	         "  con         BarcodeManager connections\n"
	         "  dev         devices managed by BarcodeManager\n\n"),
	          prog_name);
}

static BMCResultCode 
do_help (BmCli *bmc, int argc, char **argv)
{
	usage ("bmcli");
	return BMC_RESULT_SUCCESS;
}

static const struct cmd {
	const char *cmd;
	BMCResultCode (*func) (BmCli *bmmc, int argc, char **argv);
} bmcli_cmds[] = {
	{ "bm",         do_barcode_manager },
	{ "con",        do_connections },
	{ "dev",        do_devices },
	{ "help",       do_help },
	{ 0 }
};

static BMCResultCode
do_cmd (BmCli *bmc, const char *argv0, int argc, char **argv)
{
	const struct cmd *c;

	for (c = bmcli_cmds; c->cmd; ++c) {
		if (matches (argv0, c->cmd) == 0)
			return c->func (bmc, argc-1, argv+1);
	}

	g_string_printf (bmc->return_text, _("Error: Object '%s' is unknown, try 'bmcli help'."), argv0);
	bmc->return_value = BMC_RESULT_ERROR_USER_INPUT;
	return bmc->return_value;
}

static BMCResultCode
parse_command_line (BmCli *bmc, int argc, char **argv)
{
	char *base;

	base = strrchr (argv[0], '/');
	if (base == NULL)
		base = argv[0];
	else
		base++;

	/* parse options */
	while (argc > 1) {
		char *opt = argv[1];
		/* '--' ends options */
		if (strcmp (opt, "--") == 0) {
			argc--; argv++;
			break;
		}
		if (opt[0] != '-')
			break;
		if (opt[1] == '-')
			opt++;
		if (matches (opt, "-terse") == 0) {
			if (bmc->print_output == BMC_PRINT_TERSE) {
				g_string_printf (bmc->return_text, _("Error: Option '--terse' is specified the second time."));
				bmc->return_value = BMC_RESULT_ERROR_USER_INPUT;
				return bmc->return_value;
			}
			else if (bmc->print_output == BMC_PRINT_PRETTY) {
				g_string_printf (bmc->return_text, _("Error: Option '--terse' is mutually exclusive with '--pretty'."));
				bmc->return_value = BMC_RESULT_ERROR_USER_INPUT;
				return bmc->return_value;
			}
			else
				bmc->print_output = BMC_PRINT_TERSE;
		} else if (matches (opt, "-pretty") == 0) {
			if (bmc->print_output == BMC_PRINT_PRETTY) {
				g_string_printf (bmc->return_text, _("Error: Option '--pretty' is specified the second time."));
				bmc->return_value = BMC_RESULT_ERROR_USER_INPUT;
				return bmc->return_value;
			}
			else if (bmc->print_output == BMC_PRINT_TERSE) {
				g_string_printf (bmc->return_text, _("Error: Option '--pretty' is mutually exclusive with '--terse'."));
				bmc->return_value = BMC_RESULT_ERROR_USER_INPUT;
				return bmc->return_value;
			}
			else
				bmc->print_output = BMC_PRINT_PRETTY;
		} else if (matches (opt, "-mode") == 0) {
			bmc->mode_specified = TRUE;
			next_arg (&argc, &argv);
			if (argc <= 1) {
		 		g_string_printf (bmc->return_text, _("Error: missing argument for '%s' option."), opt);
				bmc->return_value = BMC_RESULT_ERROR_USER_INPUT;
				return bmc->return_value;
			}
			if (!strcmp (argv[1], "tabular"))
				bmc->multiline_output = FALSE;
			else if (!strcmp (argv[1], "multiline"))
				bmc->multiline_output = TRUE;
			else {
		 		g_string_printf (bmc->return_text, _("Error: '%s' is not valid argument for '%s' option."), argv[1], opt);
				bmc->return_value = BMC_RESULT_ERROR_USER_INPUT;
				return bmc->return_value;
			}
		} else if (matches (opt, "-escape") == 0) {
			next_arg (&argc, &argv);
			if (argc <= 1) {
		 		g_string_printf (bmc->return_text, _("Error: missing argument for '%s' option."), opt);
				bmc->return_value = BMC_RESULT_ERROR_USER_INPUT;
				return bmc->return_value;
			}
			if (!strcmp (argv[1], "yes"))
				bmc->escape_values = TRUE;
			else if (!strcmp (argv[1], "no"))
				bmc->escape_values = FALSE;
			else {
		 		g_string_printf (bmc->return_text, _("Error: '%s' is not valid argument for '%s' option."), argv[1], opt);
				bmc->return_value = BMC_RESULT_ERROR_USER_INPUT;
				return bmc->return_value;
			}
		} else if (matches (opt, "-fields") == 0) {
			next_arg (&argc, &argv);
			if (argc <= 1) {
		 		g_string_printf (bmc->return_text, _("Error: fields for '%s' options are missing."), opt);
				bmc->return_value = BMC_RESULT_ERROR_USER_INPUT;
				return bmc->return_value;
			}
			bmc->required_fields = g_strdup (argv[1]);
		} else if (matches (opt, "-version") == 0) {
			printf (_("bmcli tool, version %s\n"), BMCLI_VERSION);
			return BMC_RESULT_SUCCESS;
		} else if (matches (opt, "-help") == 0) {
			usage (base);
			return BMC_RESULT_SUCCESS;
		} else {
			g_string_printf (bmc->return_text, _("Error: Option '%s' is unknown, try 'bmcli -help'."), opt);
			bmc->return_value = BMC_RESULT_ERROR_USER_INPUT;
			return bmc->return_value;
		}
		argc--;
		argv++;
	}

	if (argc > 1)
		return do_cmd (bmc, argv[1], argc-1, argv+1);

	usage (base);
	return bmc->return_value;
}

static void
signal_handler (int signo)
{
	if (signo == SIGINT || signo == SIGTERM) {
		g_message (_("Caught signal %d, shutting down..."), signo);
		g_main_loop_quit (loop);
	}
}

static void
setup_signals (void)
{
	struct sigaction action;
	sigset_t mask;

	sigemptyset (&mask);
	action.sa_handler = signal_handler;
	action.sa_mask = mask;
	action.sa_flags = 0;
	sigaction (SIGTERM,  &action, NULL);
	sigaction (SIGINT,  &action, NULL);
}

static BMClient *
bmc_get_client (BmCli *bmc)
{
	if (!bmc->client) {
		bmc->client = bm_client_new ();
		if (!bmc->client) {
			g_critical (_("Error: Could not create BMClient object."));
			exit (BMC_RESULT_ERROR_UNKNOWN);
		}
	}

	return bmc->client;
}

/* Initialize BmCli structure - set default values */
static void
bmc_init (BmCli *bmc)
{
	bmc->client = NULL;
	bmc->get_client = &bmc_get_client;

	bmc->return_value = BMC_RESULT_SUCCESS;
	bmc->return_text = g_string_new (_("Success"));

	bmc->timeout = 10;

	bmc->system_settings = NULL;
	bmc->user_settings = NULL;

	bmc->system_settings_running = FALSE;
	bmc->user_settings_running = FALSE;

	bmc->system_connections = NULL;
	bmc->user_connections = NULL;

	bmc->should_wait = FALSE;
	bmc->nowait_flag = TRUE;
	bmc->print_output = BMC_PRINT_NORMAL;
	bmc->multiline_output = FALSE;
	bmc->mode_specified = FALSE;
	bmc->escape_values = TRUE;
	bmc->required_fields = NULL;
	bmc->allowed_fields = NULL;
	memset (&bmc->print_fields, '\0', sizeof (BmcPrintFields));
}

static void
bmc_cleanup (BmCli *bmc)
{
	if (bmc->client) g_object_unref (bmc->client);

	g_string_free (bmc->return_text, TRUE);

	if (bmc->system_settings) g_object_unref (bmc->system_settings);
	if (bmc->user_settings) g_object_unref (bmc->user_settings);

	g_slist_free (bmc->system_connections);
	g_slist_free (bmc->user_connections);

	g_free (bmc->required_fields);
	if (bmc->print_fields.indices)
		g_array_free (bmc->print_fields.indices, TRUE);
}

static gboolean
start (gpointer data)
{
	ArgsInfo *info = (ArgsInfo *) data;
	info->bmc->return_value = parse_command_line (info->bmc, info->argc, info->argv);

	if (!info->bmc->should_wait)
		g_main_loop_quit (loop);

	return FALSE;
}


int
main (int argc, char *argv[])
{
        BmCli bmc;
	ArgsInfo args_info = { &bmc, argc, argv };

	/* Set locale to use environment variables */
	setlocale (LC_ALL, "");

#ifdef GETTEXT_PACKAGE
	/* Set i18n stuff */
	bindtextdomain (GETTEXT_PACKAGE, BMCLI_LOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);
#endif

	g_type_init ();

	bmc_init (&bmc);
	g_idle_add (start, &args_info);

	loop = g_main_loop_new (NULL, FALSE);  /* create main loop */
	setup_signals ();                      /* setup UNIX signals */
	g_main_loop_run (loop);                /* run main loop */

	/* Print result descripting text */
	if (bmc.return_value != BMC_RESULT_SUCCESS) {
		fprintf (stderr, "%s\n", bmc.return_text->str);
	}

	g_main_loop_unref (loop);
	bmc_cleanup (&bmc);

	return bmc.return_value;
}
